src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Cohost/CohostDocumentCompletionEndpointTest.cs

```diff
@@ -0,0 +1,508 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.AspNetCore.Razor.Test.Common;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.Razor.Settings;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.Razor.Settings;
+using Microsoft.VisualStudio.Razor.Snippets;
+using Xunit;
+using Xunit.Abstractions;
+using RoslynCompletionParams = Roslyn.LanguageServer.Protocol.CompletionParams;
+using RoslynCompletionTriggerKind = Roslyn.LanguageServer.Protocol.CompletionTriggerKind;
+using RoslynLspExtensions = Roslyn.LanguageServer.Protocol.RoslynLspExtensions;
+using RoslynTextDocumentIdentifier = Roslyn.LanguageServer.Protocol.TextDocumentIdentifier;
+using RoslynVSInternalCompletionContext = Roslyn.LanguageServer.Protocol.VSInternalCompletionContext;
+using RoslynVSInternalCompletionInvokeKind = Roslyn.LanguageServer.Protocol.VSInternalCompletionInvokeKind;
+
+namespace Microsoft.VisualStudio.Razor.LanguageClient.Cohost;
+
+public class CohostDocumentCompletionEndpointTest(ITestOutputHelper testOutputHelper) : CohostEndpointTestBase(testOutputHelper)
+{
+    [Fact]
+    public async Task CSharpClassesAtTransition()
+    {
+        await VerifyCompletionListAsync(
+            input: $"""
+                This is a Razor document.
+
+                <div>@$$</div>
+
+                The end.
+                """,
+             completionContext: new RoslynVSInternalCompletionContext()
+             {
+                 InvokeKind = RoslynVSInternalCompletionInvokeKind.Typing,
+                 TriggerCharacter = "@",
+                 TriggerKind = RoslynCompletionTriggerKind.TriggerCharacter
+             },
+             expectedItemLabels: ["char", "DateTime", "Exception"],
+             expectedItemCount: 996);
```

Comment:
Is validating the count valuable? Seems pretty fragile.

Reply:
As I mentioned in one of the comments, Roslyn doesn't seem to always return all items initially (possibly related to our async code generation or something like that?). Waiting for expected number of items ensures we are waiting for everything to be ready (not sure if there is a better event to wait for?). I agree it may end up being fragile. If it is, we could switch from checking for exact match to checking that returned count is greater than the baseline number. I'd like to leave it as is for now and see if it causes issues, in which case we can modify the check.

Reply:
Removed item count per our offline online conversation.